src/Native/Runtime/arm/WriteBarriers.S

```diff
@@ -1,68 +1,378 @@
+          ldr          r3, =C_FUNC(g_lowest_address)
+          ldr          r3, [r3]
+          cmp          r0, r3
+          blo          LOCAL_LABEL(RhpByRefAssignRef_NotInHeap)
+          ldr          r3, =C_FUNC(g_highest_address)
+          ldr          r3, [r3]
+          cmp          r0, r3
+          bhi          LOCAL_LABEL(RhpByRefAssignRef_NotInHeap)
 
-// WARNING: Code in EHHelpers.cpp makes assumptions about write barrier code, in particular:
-// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpCheckedLockCmpXchgAVLocation
-// - Function "UnwindWriteBarrierToCaller" assumes the stack contains just the pushed return address
-LEAF_ENTRY RhpCheckedLockCmpXchg, _TEXT
-ALTERNATE_ENTRY RhpCheckedLockCmpXchgAVLocation
-LEAF_END RhpCheckedLockCmpXchg, _TEXT
+          // Update the shadow copy of the heap with the same value just written to the same heap. (A no-op unless
+          // we're in a debug build and write barrier checking has been enabled).
+          UPDATE_GC_SHADOW BASENAME, r2, r0
 
+          // If the reference is to an object that's not in an ephemeral generation we have no need to track it
+          // (since the object won't be collected or moved by an ephemeral collection).
+          ldr          r3, =C_FUNC(g_ephemeral_low)
+          ldr          r3, [r3]
+          cmp          r2, r3
+          blo          LOCAL_LABEL(RhpByRefAssignRef_NotInHeap)
+          ldr          r3, =C_FUNC(g_ephemeral_high)
+          ldr          r3, [r3]
+          cmp          r2, r3
+          bhi          LOCAL_LABEL(RhpByRefAssignRef_NotInHeap)
 
-// WARNING: Code in EHHelpers.cpp makes assumptions about write barrier code, in particular:
-// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpCheckedXchgAVLocation
-// - Function "UnwindWriteBarrierToCaller" assumes the stack contains just the pushed return address
-LEAF_ENTRY RhpCheckedXchg, _TEXT
+          // move current r0 value into r2 and then increment the pointers
+          mov          r2, r0
+          add          r1, #4
+          add          r0, #4
 
-// Setup rax with the new object for the exchange, that way it will automatically hold the correct result
-// afterwards and we can leave rdx unaltered ready for the GC write barrier below.
-ALTERNATE_ENTRY RhpCheckedXchgAVLocation
-LEAF_END RhpCheckedXchg, _TEXT
+          // We have a location on the GC heap being updated with a reference to an ephemeral object so we must
+          // track this write. The location address is translated into an offset in the card table bitmap. We set
+          // an entire byte in the card table since it's quicker than messing around with bitmasks and we only write
+          // the byte if it hasn't already been done since writes are expensive and impact scaling.
+          lsr           r2, #LOG2_CLUMP_SIZE
```

Comment:
This shift should be done as part of address mode - look for `add     r12, r12, $DESTREG, lsr #10` in the Windows version